ecosystem/sep-0010.md

```diff
@@ -250,6 +257,25 @@ A server may choose to issue JWTs for less than all thresholds and based on any
 
 A server that needs to support validating accounts that do not exist can require a signature of the master key of the account address for accounts that do not exist.
 
+### Verifying Client Application Identity
+
+A server may want to attribute each HTTP request to a specific client application. For example, a server and client could make a deal to elimate fees for the users of that client. SEP-10 offers an optional variation to the flow outlined in the [Abstract](#abstract) to enable this use case. 
+
+In order to use this optional feature, the organization that owns the client application must host a [SEP-1 stellar.toml](sep-0001.md) file containing a `SIGNING_KEY` attribute. This allows the SEP-10 server to verify that the challenge sent by a client is signed by the `SIGNING_KEY`, proving that the organization takes ownership of the client application. The SEP-10 server can now attribute requests made with the resulting JWT to the organization that signed the challenge.
+
+1. The client makes the initial `GET <WEB_AUTH_ENDPOINT>` request with an added `?client_domain=example.com` parameter and value
+1. The SEP-10 server generates the challenge transaction with an additional Manage Data operation containing the `<client_domain> attribution>` passed 
```

Comment:
We also should specify here what the source account should be.

Reply:
Why is the source account important? The SEP currently just requests that source accounts not be null.

Reply:
This transaction should follow the basic rules of Stellar transactions. The signature for an operation needs to be the signer of the source account of the operation. If the source account is the server account then the server should sign for it.

The rules about additional operations signed by the server is just to allow a server to include any additional information itself.

Reply:
Hmm ok I see where you're coming from, but I don't think we should try to apply these rules to challenge transactions, and I would argue that we're already breaking these rules with SEP-10 3.1's changes.

In a payment operation from account A to account B, I can have a transaction source account C, as long as A is the source account on the payment operation. Account A needs to sign the transaction along with account C because A's funds are being used and C's sequence number is being used. That all makes sense.

But in challenge transactions, how to go about applying these rules is not as clear. In the first Manage Data operation, the client's account is the source account, but the data is the home domain of the server's TOML file. You could argue that the client needs to verify the home domain included and signs the transaction to "approve" of the home domain -- this is why we updated to SEP-10 3.0.

But before SEP-10 3.0, the client's signature's only use was proving to the server the client controlled the Stellar account specified. My point here is that unlike in payment operations, the Manage Data operation was not the reason the client needed to sign the transaction.

My point is further reinforced by SEP-10 3.1's changes. The added operation uses the server's account as the source account. Why is the server the source account for the `web_auth_domain` but not the `home_domain`? Shouldn't the client account be the source account on this new op, since the reason we're introducing it is for the client to verify that the challenge originated from that domain?

All this is to say that instead trying to apply the signing rules of normal "submittable" transactions to challenge transactions, we should acknowledge that challenge transaction signatures prove a specific claim.
- the server's signature proves possession of the secret for the `SIGNING_KEY`'s on the `home_domain`'s TOML
- the user's signature proves the client is in possession of the secret for the Stellar account being authenticated
- the client's signature proves the client is in possession of the secret for the `SIGNING_KEY` on the `client_domain`'s TOML, verifying that the client wants to be attributed to the authentication request

Reply:
The first operation needs to be an exception. It is how it is and future operations we add don't need to conform to it. I think these rules should be applied to how we define the subsequent operations.

I think the source account of an operation should otherwise indicate who is signing off on that claim to make it clear if the client is signing it, a third party, or simply the server. This is why SEP-2.1 included a statement that "if the transaction has other operations they are Manage Data operations that all have their source accounts set to the the server's account." The reason it makes that statement is if there are operations included that the client doesn't recognize, and they're not the servers, the client shouldn't sign in case the server thinks the client is signing off on them when it doesn't intend to.

Another issue is if the source account isn't included then the server is required to either store the information about who else is supposed to sign, or look that information up fresh when receiving the challenge.

Sticking it in the source account field makes it really clear who is signing for that operation.

What's the motivation for going a different route with this?

Reply:
I think the notion of _signing for an operation_ doesn't cleanly apply to special non-submittable transactions like challenges. For challenge transactions, both client and server sign to prove each party is who they say they are, not because a particular operation requires either party's signature.

But in this instance we agree that the source account is the server account for the attribution Manage Data op.

Reply:
I think the source account for the attribution op should be the party signing to prove that claim, which is the third party. The SEP-10 server isn't proving that attribute.

Reply:
Oh I see, this requires the SEP-10 server to make request to `<client_domain>/.well-known/stellar.toml` so it can extract the `SIGNING_KEY`. But maybe this isn't a concern because Anchors could only attribute challenges to known `client_domains` for which they have an internal mapping to the client server's `SIGNING_KEY`.

Reply:
Yup. The server needs the signing key of the third party anyway, to verify the signature on the challenge tx.

Reply:
I'm going to add a point about the above ^